cmd/metrics.go

```diff
@@ -0,0 +1,173 @@
+		},
+		[]string{"project", "topics", "ref", "stage", "job_name"},
+	)
+
+	jobRunCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "gitlab_ci_pipeline_job_run_count",
+			Help: "GitLab CI pipeline job run count",
+		},
+		[]string{"project", "topics", "ref", "stage", "job_name"},
+	)
+
+	lastRunID = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "gitlab_ci_pipeline_last_run_id",
+			Help: "ID of the most recent pipeline",
+		},
+		[]string{"project", "topics", "ref"},
+	)
+
+	lastRunStatus = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "gitlab_ci_pipeline_last_run_status",
+			Help: "Status of the most recent pipeline",
+		},
+		[]string{"project", "topics", "ref", "status"},
+	)
+
+	runCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "gitlab_ci_pipeline_run_count",
+			Help: "GitLab CI pipeline run count",
+		},
+		[]string{"project", "topics", "ref"},
+	)
+
+	timeSinceLastRun = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "gitlab_ci_pipeline_time_since_last_run_seconds",
+			Help: "Elapsed time since most recent GitLab CI pipeline run.",
+		},
+		[]string{"project", "topics", "ref"},
+	)
+
+	pipelineVariables = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "gitlab_ci_pipeline_run_count_with_variable",
+			Help: "Count of pipelines with variables",
+		},
+		[]string{"project", "ref", "pipeline_variables"},
+	)
+)
+
+var (
+	registry       = prometheus.NewRegistry()
+	defaultMetrics = []prometheus.Collector{
+		coverage,
+		lastRunDuration,
+		lastRunJobDuration,
+		lastRunJobStatus,
+		lastRunJobArtifactSize,
+		timeSinceLastJobRun,
+		jobRunCount,
+		lastRunID,
+		lastRunStatus,
+		runCount,
+		timeSinceLastRun,
+		pipelineVariables,
+	}
+)
+
+func registerMetricOn(registry *prometheus.Registry, metrics ...prometheus.Collector) {
+	for _, m := range metrics {
+		if err := registry.Register(m); err != nil {
+			panic(fmt.Errorf("could not add provided metric '%v' to the Prometheus registry: %v", m, err))
+		}
+	}
+}
+
+func emitStatusMetric(metric *prometheus.GaugeVec, labelValues []string, statuses []string, status string, sparseMetrics bool) {
+	// Moved into separate function to reduce cyclomatic complexity
+	// List of available statuses from the API spec
+	// ref: https://docs.gitlab.com/ee/api/jobs.html#list-pipeline-jobs
+	for _, s := range statuses {
+		args := append(labelValues, s)
+		if s == status {
+			metric.WithLabelValues(args...).Set(1)
+		} else {
+			if sparseMetrics {
+				metric.DeleteLabelValues(args...)
+			} else {
+				metric.WithLabelValues(args...).Set(0)
+			}
+		}
+	}
+}
+
+type pipelineVarsFetchOp func(interface{}, int, ...gitlab.RequestOptionFunc) ([]*gitlab.PipelineVariable, *gitlab.Response, error)
+
+func emitPipelineVariablesMetric(c *Client, metric *prometheus.GaugeVec, projectName, ref string, projectID int, pipelineID int, fetch pipelineVarsFetchOp) error {
```

Comment:
This could lead to a very large amount of data being fetched. I believe that most of the time, the user will not need it or only a few of them. It could be nice to add a regex to filter-in the CI variable that the user want to get on a per project/wildcard basis. What do you think?

Reply:
ok, can be an improvement.
would you like to gate in a config the read of such metrics?

Reply:
yes that would be great! 🙏 🙇

Reply:
awesome, whilst at it I was thinking that actually instead of duplicating `gitlab_ci_pipeline_run_count_with_variable` since `gitlab_ci_pipeline_run_count` already exist we could add the labels there (provided config flag is given and filtered with regex)!
WDYT?

Reply:
implemented the gating and filtering, probably we can keep separated for now

Reply:
I agree let's add it straight into `gitlab_ci_pipeline_run_count` 👍

Reply:
I'll see what's the best way of doing it